juju/client/connection.py

```diff
@@ -74,17 +76,28 @@ def _get_ssl(self, cert=None):
         kw['loop'] = self.loop
         self.addr = url
         self.ws = await websockets.connect(url, **kw)
+        self.loop.create_task(self.receiver())
         log.info("Driver connected to juju %s", url)
         return self
 
     async def close(self):
         await self.ws.close()
 
-    async def recv(self):
-        result = await self.ws.recv()
-        if result is not None:
-            result = json.loads(result)
-        return result
+    async def recv(self, request_id):
+        if not self.is_open:
+            raise JujuError('Connection closed')
```

Comment:
This only catches the case where we're making the request after self.is_open is false.

It doesn't handle the case where you're in the middle of waiting for a message on the queue. I'm not sure that there's a great way to deal w/ the latter, however ...